File: src/main/history-store.js
'use strict';

const fs = require('fs');
const path = require('path');

const MAX_ENTRIES = 20000;

/**
 * Browsing history is always encrypted at rest using the OS keychain-backed
 * `safeStorage` API (Keychain / DPAPI / libsecret), independent of the user's
 * PIN. The PIN is a separate, in-app gate: it doesn't derive the encryption
 * key, so background history logging keeps working without prompting for a
 * PIN on every page load - the PIN only gates *viewing* history in the UI.
 */
class HistoryStore {
  constructor(userDataDir, safeStorage) {
    this.filePath = path.join(userDataDir, 'history.enc');
    this.safeStorage = safeStorage;
    this.entries = this._load();
  }

  _load() {
    if (!fs.existsSync(this.filePath)) return [];
    try {
      const encrypted = fs.readFileSync(this.filePath);
      if (!this.safeStorage.isEncryptionAvailable()) return [];
      const decrypted = this.safeStorage.decryptString(encrypted);
      return JSON.parse(decrypted);
    } catch (err) {
      return [];
    }
  }

  _persist() {
    if (!this.safeStorage.isEncryptionAvailable()) return;
    const encrypted = this.safeStorage.encryptString(JSON.stringify(this.entries));
    fs.writeFileSync(this.filePath, encrypted);
  }

  record(url, title) {
    if (!url || url.startsWith('dago://')) return;
    this.entries.push({ url, title: title || url, timestamp: Date.now() });
    if (this.entries.length > MAX_ENTRIES) this.entries.shift();
    this._persist();
  }

  list() {
    return [...this.entries].reverse();
  }

  clear() {
    this.entries = [];
    this._persist();
  }

  removeEntry(index) {
    const reversedIdx = this.entries.length - 1 - index;
    if (reversedIdx >= 0 && reversedIdx < this.entries.length) {
      this.entries.splice(reversedIdx, 1);
      this._persist();
    }
  }
}

module.exports = { HistoryStore };

File: android/app/src/main/java/org/dago/browser/HistoryStore.kt
package org.dago.browser

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONArray
import org.json.JSONObject

private const val MAX_ENTRIES = 20_000

/**
 * Mirrors the desktop app's history-store.js: history is always encrypted
 * at rest - there, via OS-keychain-backed `safeStorage`; here, via the
 * Android Keystore-backed `EncryptedSharedPreferences` - independent of the
 * PIN, so recording keeps working in the background without prompting.
 * [PinStore] separately gates *viewing* history in the UI.
 */
class HistoryStore(context: Context) {
    data class Entry(val url: String, val title: String, val timestamp: Long)

    private val prefs = run {
        val masterKey = MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
        EncryptedSharedPreferences.create(
            context,
            "dago_history",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    fun record(url: String, title: String) {
        if (url.isBlank() || url == "about:blank") return
        val entries = loadRaw()
        entries.put(
            JSONObject()
                .put("url", url)
                .put("title", title)
                .put("timestamp", System.currentTimeMillis())
        )
        while (entries.length() > MAX_ENTRIES) entries.remove(0)
        persist(entries)
    }

    /** Newest first. */
    fun list(): List<Entry> {
        val raw = loadRaw()
        val result = mutableListOf<Entry>()
        for (i in 0 until raw.length()) {
            val obj = raw.getJSONObject(i)
            result.add(Entry(obj.getString("url"), obj.getString("title"), obj.getLong("timestamp")))
        }
        return result.reversed()
    }

    fun removeEntry(displayIndex: Int) {
        val raw = loadRaw()
        val reversedIdx = raw.length() - 1 - displayIndex
        if (reversedIdx !in 0 until raw.length()) return
        val rebuilt = JSONArray()
        for (i in 0 until raw.length()) if (i != reversedIdx) rebuilt.put(raw.get(i))
        persist(rebuilt)
    }

    fun clear() {
        prefs.edit().remove("entries").apply()
    }

    private fun loadRaw(): JSONArray {
        val raw = prefs.getString("entries", null) ?: return JSONArray()
        return runCatching { JSONArray(raw) }.getOrDefault(JSONArray())
    }

    private fun persist(entries: JSONArray) {
        prefs.edit().putString("entries", entries.toString()).apply()
    }
}
